src/lib/snarky/src/snark0.ml

```diff
@@ -914,30 +916,28 @@ module Make_basic (Backend : Backend_intf.S) = struct
         match t with
         | Pure x -> (s, x)
         | With_constraint_system (f, k) ->
-            ignore (Option.map ~f system) ;
-            go stack k handler s
+            Option.iter ~f system ; go stack k handler s
         | With_label (lab, t, k) ->
             let s', y = go (lab :: stack) t handler s in
             go stack (k y) handler s'
         | As_prover (x, k) ->
-            let s', _ = run_as_prover (Some x) s in
+            let s', (_ : unit option) = run_as_prover (Some x) s in
             go stack k handler s'
         | Add_constraint (c, t) ->
-            ignore
-              (Option.map system ~f:(fun system ->
-                   (* NOTE: If s is None, we aren't evaluating and shouldn't consider constraints here. *)
-                   if Option.is_some s && not (Constraint.eval c get_value)
-                   then
-                     failwithf "Constraint unsatisfied:\n%s\n%s\n"
-                       (Constraint.annotation c)
-                       (Constraint.stack_to_string stack)
-                       () ;
-                   Constraint.add ~stack c system )) ;
+            Option.iter system ~f:(fun system ->
```

Comment:
I think that checking constraints is orthogonal to whether or not we add them to the constraint system. E.g., in run_and_check we don't have a constraint system but we still need to be checking constraints.

Reply:
It should be adding the constraints either way (as long as there is a `system` to add them to). The `if eval_constraints && ... then` only covers the `failwithf`.